plugins/registry.js

```diff
@@ -0,0 +1,207 @@
+            name: 'MobileChannelHeaderButton',
+            data: {
+                id,
+                pluginId: this.id,
+                buttonComponent,
+                dropdownComponent,
+            },
+        });
+
+        return id;
+    }
+
+    // Add a button to the mobile channel header. If there are more than one buttons registered by any
+    // plugin, buttons will instead be added to the channel header dropdown.
+    // Accepts the following:
+    // - icon - JSX element to use as the button's icon
+    // - action - a function called when the button is clicked
+    // - dropdown_text - string or JSX element shown for the dropdown button description
+    registerMobileChannelHeaderButtonAction = (icon, action, dropdownText) => {
+        const id = generateId();
+
+        store.dispatch({
+            type: ActionTypes.RECEIVED_PLUGIN_COMPONENT,
+            name: 'MobileChannelHeaderButton',
+            data: {
+                id,
+                pluginId: this.id,
+                icon,
+                action,
+                dropdownText,
+            },
+        });
+
+        return id;
+    }
+
+    // Register a component to render a custom body for posts with a specific type.
+    // Custom post types must be prefixed with 'custom_'.
+    // Accepts a string type and a component.
+    // Returns a unique identifier.
+    registerPostTypeComponent = (type, component) => {
+        const id = generateId();
+
+        store.dispatch({
+            type: ActionTypes.RECEIVED_PLUGIN_POST_COMPONENT,
+            data: {
+                id,
+                pluginId: this.id,
+                type,
+                component,
+            },
+        });
+
+        return id;
+    }
+
+    // Register a main menu list item by providing some text and an action function.
+    // Accepts the following:
+    // - text - A string or JSX element to display in the menu
+    // - action - A function to trigger when component is clicked on
+    // - mobileIcon - An icon to display in the menu in mobile view
+    // Returns a unique identifier.
+    registerMainMenuAction = (text, action, mobileIcon) => {
+        const id = generateId();
+
+        store.dispatch({
+            type: ActionTypes.RECEIVED_PLUGIN_COMPONENT,
+            name: 'MainMenu',
+            data: {
+                id,
+                pluginId: this.id,
+                text,
+                action,
+                mobileIcon,
+            },
+        });
+
+        return id;
+    }
+
+    // Deregister a component using the unique identifier returned after registration.
+    // Accepts a string id.
+    // Returns undefined in all cases.
+    deregisterComponent = (componentId) => {
+        store.dispatch({
+            type: ActionTypes.REMOVED_PLUGIN_COMPONENT,
+            id: componentId,
+        });
+    }
+
+    // Register a reducer against the Redux store. It will accessible in redux state
+    // under "state['plugins-<yourpluginid>']"
```

Comment:
It would have to be something other than`state.plugins` as that already exists and we don't want them to accidentally clobber one of our web app reducers. I'd also have to add support for that to the reducer registry (or write some workaround code here) since it currently only allows registering reducers at the top level